File: youkon/YouKon/YouKon/views/QuickConvertCard.swift
//
//  QuickConvertCard.swift
//  YouKon
//
//  Created by Eliott Radcliffe on 6/20/23.
//  Copyright © 2023 DC Simulation Studio. All rights reserved.
//

import SwiftUI
import shared

/// The quick convert card is shown at the top of the screen, and lets the user convert a single measurement value from any unit to any equivalent unit.
struct QuickConvertCard: View {
    @ObservedObject var vc = QuickConvertCardController()
    
    var body: some View {
        GroupBox(
            label: Text("Quick Convert")
        ) {
            VStack(alignment: .leading, spacing: 16) {
                userInputRow
                convertedText
            }
        }
        .frame(width: 360)
    }
    
    /// The row with the text field on the left, and menu buttons for "From" and "To" units
    @ViewBuilder
    var userInputRow: some View {
        HStack(spacing: 8) {
            textField
            fromDropdown
            toDropdown
        }
    }
    
    /// The field that takes the user input on the numeric value of the measurement
    @ViewBuilder
    private var textField: some View {
        MeasurementTextField(
            measurement: $vc.measurement,
            updateMeasurement: {
                vc.setConvertedText()
            }
        )
    }
    
    /// Selection for which type of unit to convert from
    @ViewBuilder
    private var fromDropdown: some View {
        UnitDropdown(
            unit: $vc.measurement.unit,
            availableUnits: $vc.allUnits,
            headerText: "From"
        ) { unit in
            vc.updateUnit(to: unit)
        }
    }
    
    /// Selection for which type of unit to convert to
    @ViewBuilder
    private var toDropdown: some View {
        UnitDropdown(
            unit: $vc.targetUnit,
            availableUnits: $vc.equivalentUnits,
            headerText: "To"
        ) { _ in
            vc.setConvertedText()
        }
    }
    
    /// The display of the measurement after conversion
    @ViewBuilder
    private var convertedText: some View {
        Text(vc.convertedText)
            .padding(.leading, 8)
            .font(.subheadline)
    }
}


struct QuickConvertCard_Previews: PreviewProvider {
    static var previews: some View {
        QuickConvertCard()
    }
}

File: youkon/YouKon/src/main/java/com/dcengineer/youkon/views/QuickConvertCard.kt
package com.dcengineer.youkon.views

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.dcengineer.youkon.viewmodels.QuickConvertCardViewModel
import com.dcengineer.youkon.viewmodels.QuickConvertViews

class QuickConvertCard(
    private val vm: QuickConvertCardViewModel = QuickConvertCardViewModel()
) {
    @Composable
    fun Body() {
        Surface(
            color = MaterialTheme.colorScheme.surface.copy(alpha = 0.4f),
            shape = MaterialTheme.shapes.medium,
            modifier = OnboardingModifier(QuickConvertViews.SURFACE)
                .width(420.dp)
        ) {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                CardLabel()
                ContentGrid()
            }
        }
    }

    /// Provides a view modifier for a colored shadow if the selected view is highlighted in the onboarding screen
    @Composable
    fun OnboardingModifier(view: QuickConvertViews): Modifier {
        return if (vm.highlightedView.value == view)
            Modifier.coloredShadow(animatedColor)
        else
            Modifier
    }

    /// The label at the top of the card
    @Composable
    private fun CardLabel() {
        Text("Quick Convert",
            fontWeight = FontWeight.SemiBold,
            color = MaterialTheme.colorScheme.onSurface,
            style = MaterialTheme.typography.titleLarge
        )
    }

    /// A 2x2 grid with from dropdown in the upper left, to dropdown in the upper right, editable value in the lower left, and converted value in the lower right
    @Composable
    private fun ContentGrid() {
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            FromDropdown(OnboardingModifier(QuickConvertViews.FROM).weight(1f))
            ToDropdown(OnboardingModifier(QuickConvertViews.TO).weight(1f))
        }
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            TextField(OnboardingModifier(QuickConvertViews.VALUE).weight(1f))
            ConvertedText(OnboardingModifier(QuickConvertViews.CONVERTED).weight(1f))
        }
    }

    /// Selection for which type of unit to convert from
    @Composable
    private fun FromDropdown(modifier: Modifier) {
        val measurement by vm.measurement.collectAsState()
        UnitDropdown(
            unit = measurement.unit,
            availableUnits = vm.allUnits,
            headerText = "From",
            modifier = modifier,
            onClick = { vm.updateUnit(it) }
        ).Body()
    }

    /// Selection for which type of unit to convert to
    @Composable
    private fun ToDropdown(modifier: Modifier) {
        UnitDropdown(
            unit = vm.targetUnit,
            availableUnits = vm.equivalentUnits,
            headerText = "To",
            modifier = modifier,
            onClick = { vm.updateTargetUnit(it) }
        ).Body()
    }

    /// The field that takes the user input on the numeric value of the measurement
    @Composable
    private fun TextField(modifier: Modifier) {
        val measurement = vm.measurement.collectAsState()
        MeasurementTextField(
            initialText = measurement.value.value.toString(),
            unitText = measurement.value.unit.shortUnit,
            modifier = modifier,
            updateMeasurement = { vm.updateValue(it) }
        )
    }

    /// The display of the measurement after conversion
    @Composable
    private fun ConvertedText(modifier: Modifier) {
        Box(
            modifier = modifier,
            contentAlignment = Alignment.CenterStart
        ) {
            Text(vm.convertedText,
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface
            )
        }
    }
}

@Preview
@Composable
fun QuickConvertCardPreview() {
    QuickConvertCard().Body()
}